### PR TITLE
Use common upgrade lib and update dependencies

### PR DESCRIPTION
**Pull request by almaslennikov**
Replace built-in driver upgrade codebase with usages of a common upgrade lib

Signed-off-by: amaslennikov <amaslennikov@nvidia.com>

### CODE REVIEW

**Comment by adrianchiris** [comment]
also need to update documentation of the upgrade flow to reflect the new flow (the state machine changed right? )

File: docs/automatic-ofed-upgrade.md [hunk 1]
```diff
@@ -26,6 +26,13 @@ spec:
       # maxParallelUpgrades indicates how many nodes can be upgraded in parallel
 	  # 0 means no limit, all nodes will be upgraded in parallel
       maxParallelUpgrades: 0
+      # describes the configuration for waiting on job completions
+      waitForCompletion:
+        # specifies a label selector for the pods to wait for completion
+        podSelector: "app=myapp"
+        # specify the length of time in seconds to wait before giving up drain, zero means infinite
```
Comment (ykulazhenkov) [inline docs/automatic-ofed-upgrade.md:33]: drain?

File: go.mod [hunk 2]
```diff
@@ -4,19 +4,20 @@ go 1.18
...
-	github.com/onsi/gomega v1.20.1
+	github.com/onsi/gomega v1.24.0
 	github.com/openshift/api v0.0.0-20210428205234-a8389931bee7
 	github.com/pkg/errors v0.9.1
-	github.com/stretchr/testify v1.8.0
+	github.com/stretchr/testify v1.8.1
 	k8s.io/api v0.25.5
 	k8s.io/apimachinery v0.25.5
 	k8s.io/client-go v0.25.5
 	k8s.io/kubectl v0.25.5
```
Comment (ykulazhenkov) [inline go.mod:19]: It look like kubectl dependency is not required anymore. Could you please do `go mod tidy`

**Review by ykulazhenkov — APPROVED** [review]

**Review by e0ne — APPROVED** [review]

File: controllers/upgrade_controller.go [hunk 3]
```diff
@@ -115,20 +117,23 @@ func (r *UpgradeReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ct
 
 // removeNodeUpgradeStateAnnotations loops over nodes in the cluster and removes upgrade.UpgradeStateAnnotation
```
Comment (adrianchiris) [inline controllers/upgrade_controller.go:118]: please update comment

File: controllers/upgrade_controller.go [hunk 4]
```diff
@@ -83,7 +84,7 @@ func (r *UpgradeReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ct
 		nicClusterPolicy.Spec.OFEDDriver.OfedUpgradePolicy == nil ||
 		!nicClusterPolicy.Spec.OFEDDriver.OfedUpgradePolicy.AutoUpgrade {
 		reqLogger.V(consts.LogLevelInfo).Info("OFED Upgrade Policy is disabled, skipping driver upgrade")
-		err = r.removeNodeUpgradeStateAnnotations(ctx)
+		err = r.removeNodeUpgradeStateLabels(ctx)
```
[the remark this answers is not in the document]
Comment (e0ne) [reply controllers/upgrade_controller.go:87]: We're aligning with GPU operator [1] https://github.com/NVIDIA/gpu-operator/blob/b588e4e1cbc7fa4e875d30b5c1a3ba046421aff0/bundle/metadata/annotations.yaml#L17 and MOFED images availability. The current assumption is: testing on the latests with assumption backward compatibility works

File: go.mod [hunk 5]
```diff
@@ -4,19 +4,19 @@ go 1.18
 
 require (
 	github.com/Masterminds/semver/v3 v3.1.1
+	github.com/NVIDIA/k8s-operator-libs v0.0.0-20230118182127-8e73980389ca
```
Comment (adrianchiris) [inline go.mod:7]: nit, just making sure `go mod tidy` was run.

Comment (almaslennikov) [reply go.mod:7]: Yes, it was. Do you have a concern about something here?

File: controllers/upgrade_controller.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (almaslennikov) [reply controllers/upgrade_controller.go:96]: Makes sense, added that

File: docs/automatic-ofed-upgrade.md [hunk 6]
```diff
@@ -46,16 +53,18 @@ spec:
...
-* `drain` is set when the node is scheduled for drain. After the drain the state is changed either to `pod-restart` or `drain-failed`
-UpgradeStateDrain = "drain"
-* `pod-restart` is set when the OFED POD on the node is scheduler for restart. After the restart state is changed to `uncordon-required`
-* `drain-failed` is set when drain on the node has failed. Manual interaction is required at this stage. See [Troubleshooting](#node-is-in-drain-failed-state) section for more details.
+* `cordon-required` is set when the node needs to be made unschedulable in preparation for driver upgrade 
+* `wait-for-jobs-required` is set on the node when we need to wait on jobs to complete until given timeout
+* `drain-required` is set when the node is scheduled for drain. After the drain the state is changed either to `pod-restart` or `drain-failed`
+* `pod-restart-required` is set when the OFED POD on the node is scheduler for restart. After the restart state is changed to `uncordon-required`
+* `validation-required` is set when validation of the new driver deployed on the node is required before moving to UpgradeStateUncordonRequired.
```
[the remark this answers is not in the document]
Comment (almaslennikov) [reply docs/automatic-ofed-upgrade.md]: PTAL at a new diagram

**Review by adrianchiris — APPROVED** [review]

**Comment by almaslennikov** [comment]
@adrianchiris sure

**Pull request merged**